### PR TITLE
Gde alme auto index

### PR DESCRIPTION
**Pull request by gde-alme**
Fix Issue 19.
Also removed index/auto_index in /redirect route

### CODE REVIEW

**Comment by gde-alme** [comment]
Should fix Issues #19 , #27 & #29

**Comment by gde-alme** [comment]
Should fix Issue #20

**Comment by gde-alme** [comment]
should fix issue #20

**Comment by gde-alme** [comment]
should fix issue #20

File: include/ServerConfig.hpp
Comment (PukieDiederik) [inline include/ServerConfig.hpp:1]: If you are trying to allign the variables I would recommend using whitespace. Tabs might not always be 4 (or it seems 8? in this case) spaces wide. For example. accepted_methods for me is not alligned with the rest.

File: src/Server.cpp [hunk 1]
```diff
@@ -8,8 +8,11 @@
 #include <fstream>
 #include <sstream>
 
-Server::Server(ServerCfg& cfg, ServerConfig& gen_cfg) :_cfg(cfg), _gen_cfg(gen_cfg) { }
-Server::Server(const Server& copy) :_cfg(copy._cfg), _gen_cfg(copy._gen_cfg) { }
+Server::Server( ServerCfg& cfg, ServerConfig& gen_cfg ) :_cfg( cfg ), _gen_cfg( gen_cfg ) {
+}
+
+Server::Server( const Server& copy ) :_cfg( copy._cfg ), _gen_cfg( copy._gen_cfg ) {
+}
```
Comment (PukieDiederik) [inline src/Server.cpp:15]: Don't really see the use these style updates if it's not consistent. (Look at the destructor right under this)

File: src/Server.cpp [hunk 2]
```diff
@@ -35,6 +38,36 @@ RouteCfg* find_route(const HttpRequest& req, std::vector<RouteCfg>& routes)
...
 
+/*
+*   @check_request_method:
+*    Checks if the requested method is accepted by the route
+*/
+bool    check_request_method(RouteCfg* route, const std::string method) {
+    for (size_t i = 0; i < route->accepted_methods.size(); i++)
+        if (method == route->accepted_methods[i])
+            return true;
+    return false;
```
Comment (PukieDiederik) [inline src/Server.cpp:49]: A method already exists for finding something in a container (or any other iterrable): https://en.cppreference.com/w/cpp/algorithm/find. I would recommend using it.

File: src/Server.cpp [hunk 3]
```diff
@@ -35,6 +38,36 @@ RouteCfg* find_route(const HttpRequest& req, std::vector<RouteCfg>& routes)
...
+}
+
+/*
+*   @index_path:
+*       If auto_index is true, returns user request
+*       If not, return predefined index
+*
+*/
+std::string	index_path(const HttpRequest& req, RouteCfg* route)
```
Comment (PukieDiederik) [inline src/Server.cpp:58]: This get's or calculates the path. (There is some other cleanup that I need to do which I will make an issue for). So maybe rename this to get_path?

File: src/Server.cpp [hunk 4]
```diff
@@ -35,6 +38,36 @@ RouteCfg* find_route(const HttpRequest& req, std::vector<RouteCfg>& routes)
...
+{
+    std::string	request = "." + route->root + route->route_path;
```
Comment (PukieDiederik) [inline src/Server.cpp:60]: Why is a `.` added in front of the root? Root should be an absolute path, not relative.

File: src/Server.cpp [hunk 5]
```diff
@@ -35,6 +38,36 @@ RouteCfg* find_route(const HttpRequest& req, std::vector<RouteCfg>& routes)
...
+
+    if ( route->auto_index ) {
+        if ( req.target() == "/" ) return request;
+        if ( route->route_path == "/") return request + req.target().substr(1, req.target().size() - 1);
+
+        return "." + route->root + req.target();
```
Comment (PukieDiederik) [inline src/Server.cpp:66]: I am having trouble understanding this, please comment with a thorough explaination

File: src/Server.cpp [hunk 6]
```diff
@@ -64,6 +97,13 @@ HttpResponse Server::handleRequest(const HttpRequest& req)
         return res;
     }
 
+    std::cout << req.method() << std::endl;
```
Comment (PukieDiederik) [inline src/Server.cpp:100]: This looks like debug, Should be removed.

File: src/ServerConfig.cpp
Comment (PukieDiederik) [inline src/ServerConfig.cpp:1]: Apparently forgot to remove this. Thank you for cleaning up!

File: src/ServerConfigChecker.cpp [hunk 7]
```diff
@@ -71,32 +83,30 @@ void	ServerConfig::checker() {
...
-						throw std::runtime_error("Error: invalid route method");
 				}
+			} else { // If not accepted methods are set, set all available
+					(*jit).accepted_methods.push_back("GET");
+    				(*jit).accepted_methods.push_back("POST");
+    				(*jit).accepted_methods.push_back("PUT");
+    				(*jit).accepted_methods.push_back("PATCH");
+    				(*jit).accepted_methods.push_back("DELETE");
+					(*jit).accepted_methods.push_back("HEAD");
+					// More to be set
```
Comment (PukieDiederik) [inline src/ServerConfigChecker.cpp:107]: These are not all methods available. A request method can be anything you want, although usually these are the only ones that are implemented on servers. This is not necessary and should be removed. More info: https://www.rfc-editor.org/rfc/rfc9110.html#name-overview

File: var/www/files/test
Comment (PukieDiederik) [inline var/www/files/test:1]: Should be removed?

**Review by PukieDiederik — CHANGES_REQUESTED** [review]
Try to keep separate issues in separate branches. It allows the reviewer to more easily understand your code and in the end allows for a better code review. I've found some issues marked in the comments. For now just implement these changes here, but in future I would appreciate it if you could separate each issue in a separate branch (and PR)

File: .vscode/settings.json
[the remark this answers is not in the document]
Comment (gde-alme) [reply .vscode/settings.json:1]: should not be included did not even notice until now, i''l just remove it

File: include/ServerConfig.hpp
Comment (gde-alme) [reply include/ServerConfig.hpp:1]: It has to do with vim / vscode settings: it will always mess up between configs

File: src/Server.cpp (hunk 1 above)
Comment (gde-alme) [reply src/Server.cpp:15]: These kinds of changes are not necessarily things I "correct", instead were made when trying to get the program to run / understand what does what.

File: server/server.cfg [hunk 8]
```diff
@@ -25,10 +26,11 @@ server {
 	route "/" {
 		auto_index
 		methods ["GET"]
+		root "html"
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply server/server.cfg:29]: These kinds of changes are not necessarily things I "correct", instead were made when trying to get the program to run / understand what does what.

File: src/Server.cpp [hunk 9]
```diff
@@ -35,6 +38,36 @@ RouteCfg* find_route(const HttpRequest& req, std::vector<RouteCfg>& routes)
     return route_match.second;
 }
 
+/*
+*   @check_request_method:
+*    Checks if the requested method is accepted by the route
+*/
+bool    check_request_method(RouteCfg* route, const std::string method) {
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply src/Server.cpp:45]: Sure

File: var/www/files/test
Comment (gde-alme) [reply var/www/files/test:1]: was to test with server/server.cfg, can and will remove it

**Comment by gde-alme** [comment]
Try to keep separate issues in separate branches. It allows the reviewer to more easily understand your code and in the end allows for a better code review. I've found some issues marked in the comments. For now just implement these changes here, but in future I would appreciate it if you could separate each issue in a separate branch (and PR)

You are right , but most of the changes in this pull request  were complementary to one another, making it dificult for me to take said changes apart to different pull requests

**Comment by gde-alme** [comment]
forgot to remove the .settings file!

**Comment by gde-alme** [comment]
Could not see some of the comments , saw them now , will look into it in a bit

File: src/Server.cpp (hunk 3 above)
Comment (gde-alme) [reply src/Server.cpp:58]: Sure

File: src/Server.cpp (hunk 2 above)
Comment (gde-alme) [reply src/Server.cpp:49]: Sure

File: src/Server.cpp (hunk 4 above)
Comment (gde-alme) [reply src/Server.cpp:61]: Because the path bust be relative since the "/var/www" folder is not an absolute path

**Review by gde-alme — COMMENTED** [review]
Covered all @PukieDiederik comments i believe

File: src/ServerConfigChecker.cpp (hunk 7 above)
Comment (PukieDiederik) [reply src/ServerConfigChecker.cpp:107]: With #19 I meant implementing `methods` directive in the config. (in the subject file: `Define a list of accepted HTTP methods for the route.`) Which should only accept the request if the method is listed in the `methods` directive. In HTTP, methods are not just a predefined list of methods (Although there are some standard methods) and making a request with the request method `ABC` is also valid (Although probably not implemented). Therefore we can not possibly add all possible methods to the list (as there is literally an infinite amount of methods). So we should check if the list is empty, and if not we should check if the method is in the list.

**Comment by gde-alme** [comment]
Last changes:
In parser, check if server root path is an absolute path which can be accessed
In Server.cpp, removed "." and if root path comes with an ending '/', remove it

**Review by edgareu1 — APPROVED** [review]

File: src/Server.cpp [hunk 10]
```diff
@@ -85,7 +129,7 @@ HttpResponse Server::handleRequest(const HttpRequest& req)
         ss << res.body().length();
         res.set_status(200, "OK");
         //TODO add MIME type
-        res.set_header("Content-Type", "text/plain");
+        res.set_header("Content-Type", "text/html");
```
Comment (PukieDiederik) [inline src/Server.cpp:132]: This should be solved in master, check this when you merge.

**Review by PukieDiederik — CHANGES_REQUESTED** [review]
There are some issues (as mentioned in the comments below this). And before resubmitting it should be merged with master (Because at the moment there are merge conflicts).

I've found that I seem to be unable to request regular files (returns a 404). And when requesting a directory (with autoindex enabled in the config) It also gives me a 404 error.

File: src/ServerUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            path = route->root + route->route_path + "/" + "index.html";
+            return 0;
+        } else if ( route->auto_index ) { // else if (auto_index on), return list of contents
+            return 1;
+        }
+    }
+
+    path = "";
+    return 2;
+}
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply src/ServerUtils.cpp:48]: I checked if it was a folder thinking It should be done like that (not autoindex if was folder).
Will change it

File: include/ServerConfig.hpp [hunk 12]
```diff
@@ -12,29 +12,32 @@
 # define COMMENT 4
 # define ERROR 5
 
+# define DIRLISTING	"./var/www/templates/dir_listing.html"
+
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply include/ServerConfig.hpp:16]: So, return a dir listing with all the files in the folder in html or text format?

File: src/ServerUtils.cpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        return 0;
+    }
+    else if ( is_directory( path ) ) {
+        std::vector<std::string>    dir_listing = list_dir( path );
+        
+        // If index.html exists in said folder, return request
+        std::vector<std::string>::iterator it = std::find(dir_listing.begin(), dir_listing.end(), "index.html");
+        if ( it != dir_listing.end() ) {
+            if (!route->route_path.empty() && route->route_path[route->route_path.size() - 1] == '/')
+                route->route_path = route->route_path.substr(0, route->route_path.size() - 1);
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply src/ServerUtils.cpp:38]: Sure, makes sence

File: src/ServerUtils.cpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+            path = route->root + route->route_path + "/" + "index.html";
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply src/ServerUtils.cpp:39]: If there is an index.html file, it returns said file, if not and auto_index is on, it returns dir listing, else returns error

File: src/Server.cpp [hunk 15]
```diff
@@ -49,8 +83,9 @@ HttpResponse Server::handleRequest(const HttpRequest& req)
         return res;
     }
 
-    // Get and check path
-    path = route->root + "/" + req.target().substr(route->route_path.length());
+    // If req not a file nor index.html found return dir listing
+    if ( get_path( req, route, path ) > 0 && !path.empty() ) return list_dir_res( res, req, path);
```
[the remark this answers is not in the document]
Comment (gde-alme) [reply src/Server.cpp:87]: You told me it should be before, but will move it to inside GET

File: src/Server.cpp (hunk 10 above)
Comment (gde-alme) [reply src/Server.cpp:132]: This was a lined added to make other pages work , not what I was working on, so it can just be deleted

**Pull request merged**